storage-proofs-update/src/vanilla.rs

```diff
@@ -0,0 +1,1205 @@
+    ))]
+    pub proof_d_new: MerkleProof<TreeDHasher, TreeDArity>,
+    #[serde(bound(
+        serialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Serialize",
+        deserialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Deserialize<'de>"
+    ))]
+    pub proof_r_new:
+        MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>,
+}
+
+// Implement `Clone` by hand because `MerkleTreeTrait` does not implement `Clone`.
+impl<TreeR> Clone for ChallengeProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    fn clone(&self) -> Self {
+        ChallengeProof {
+            proof_r_old: self.proof_r_old.clone(),
+            proof_d_new: self.proof_d_new.clone(),
+            proof_r_new: self.proof_r_new.clone(),
+        }
+    }
+}
+
+impl<TreeR> ChallengeProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    pub fn verify_merkle_proofs(
+        &self,
+        c: usize,
+        comm_r_last_old: &TreeRDomain,
+        comm_d_new: &TreeDDomain,
+        comm_r_last_new: &TreeRDomain,
+    ) -> bool {
+        self.proof_r_old.path_index() == c
+            && self.proof_d_new.path_index() == c
+            && self.proof_r_new.path_index() == c
+            && self.proof_r_old.root() == *comm_r_last_old
+            && self.proof_d_new.root() == *comm_d_new
+            && self.proof_r_new.root() == *comm_r_last_new
+            && self.proof_r_old.verify()
+            && self.proof_d_new.verify()
+            && self.proof_r_new.verify()
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PartitionProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    pub apex_leafs: Vec<TreeDDomain>,
+    #[serde(bound(
+        serialize = "ChallengeProof<TreeR>: Serialize",
+        deserialize = "ChallengeProof<TreeR>: Deserialize<'de>"
+    ))]
+    pub challenge_proofs: Vec<ChallengeProof<TreeR>>,
+}
+
+// Implement `Clone` by hand because `MerkleTreeTrait` does not implement `Clone`.
+impl<TreeR> Clone for PartitionProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    fn clone(&self) -> Self {
+        PartitionProof {
+            apex_leafs: self.apex_leafs.clone(),
+            challenge_proofs: self.challenge_proofs.clone(),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(clippy::upper_case_acronyms)]
+pub struct EmptySectorUpdate<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    _tree_r: PhantomData<TreeR>,
+}
+
+impl<'a, TreeR> ProofScheme<'a> for EmptySectorUpdate<TreeR>
+where
+    TreeR: 'static + MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    type SetupParams = SetupParams;
+    type PublicParams = PublicParams;
+    type PublicInputs = PublicInputs;
+    type PrivateInputs = PrivateInputs;
+    type Proof = PartitionProof<TreeR>;
```

Comment:
`PartitionProof` is a bit odd in naming here, could you explain the reasoning?

Reply:
I guess this is just preference, but I thought that it clarified the naming.

I chose the name to differentiate it from a `ChallengeProof` (which compose a partition proof) and from a `Proof`/`VanillaProof`, which (imho) is an ambiguous name that doesn't tell you whether the proof was generated for a single or multiple partition.
```rust
impl ProofScheme for EmptySectorUpdate {
    type Proof = PartitionProof;

    fn prove(...) -> Self::Proof;
}
```
makes it explicit that `prove` is called for each partition. Also, `PartitionProof` makes it clear that a single proof should be fed as the input to a single partition's circuit.

Reply:
Makes sense